images/virtualization-artifact/pkg/controller/vd/vd_reconciler.go

```diff
@@ -0,0 +1,252 @@
+/*
+Copyright 2024 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"reflect"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/deckhouse/virtualization-controller/pkg/controller/service"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+)
+
+type Handler interface {
+	Handle(ctx context.Context, vd *virtv2.VirtualDisk) (reconcile.Result, error)
+}
+
+type Reconciler struct {
+	handlers []Handler
+	client   client.Client
+}
+
+func NewReconciler(client client.Client, handlers ...Handler) *Reconciler {
+	return &Reconciler{
+		client:   client,
+		handlers: handlers,
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	vd := service.NewResource(req.NamespacedName, r.client, r.factory, r.statusGetter)
+
+	err := vd.Fetch(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if vd.IsEmpty() {
+		return reconcile.Result{}, nil
+	}
+
+	var requeue bool
+
+	slog.Info("Start")
+
+	var handlerErrs []error
+
+	for _, h := range r.handlers {
+		var res reconcile.Result
+		slog.Info("Handle... " + reflect.TypeOf(h).Elem().Name())
```

Comment:
I think using reflect for logging is not a good idea. Can you add the method Name to the handlers?